crates/p2p-impl/src/tests/common.rs

```diff
@@ -0,0 +1,376 @@
+//! Helper functions for the P2P tests.
+
+use std::{sync::Arc, time::Duration};
+
+use anyhow::bail;
+use bitcoin::{hashes::sha256, OutPoint, XOnlyPublicKey};
+use futures::future::join_all;
+use libp2p::{
+    build_multiaddr,
+    identity::{secp256k1::Keypair as SecpKeypair, Keypair},
+    Multiaddr, PeerId,
+};
+use musig2::{PartialSignature, PubNonce};
+use strata_bridge_test_utils::{
+    musig2::{generate_partial_signature, generate_pubnonce},
+    prelude::generate_keypair,
+};
+use strata_p2p::{
+    commands::{Command, UnsignedPublishMessage},
+    events::Event,
+    swarm::{self, handle::P2PHandle, P2PConfig, P2P},
+};
+use strata_p2p_db::sled::AsyncDB;
+use strata_p2p_types::{
+    OperatorPubKey, Scope, SessionId, StakeChainId, StakeData, Wots256PublicKey,
+};
+use strata_p2p_wire::p2p::v1::{GossipsubMsg, UnsignedGossipsubMsg};
+use threadpool::ThreadPool;
+use tokio_util::{sync::CancellationToken, task::TaskTracker};
+use tracing::{info, trace};
+
+use crate::{constants::DEFAULT_NUM_THREADS, Configuration};
+
+/// Auxiliary structure to control operators from outside.
+pub(crate) struct OperatorHandle {
+    pub(crate) handle: P2PHandle,
+    pub(crate) peer_id: PeerId,
+    pub(crate) kp: SecpKeypair,
+    pub(crate) db: AsyncDB,
+}
+
+/// Represents an operator with its associated P2P instance, handle, keypair, and database.
+pub(crate) struct Operator {
+    pub(crate) p2p: P2P<AsyncDB>,
+    pub(crate) handle: P2PHandle,
+    pub(crate) kp: SecpKeypair,
+    pub(crate) db: AsyncDB,
+}
+
+impl Operator {
+    /// Creates a new operator instance.
+    pub(crate) fn new(
+        keypair: SecpKeypair,
+        allowlist: Vec<PeerId>,
+        connect_to: Vec<Multiaddr>,
+        local_addr: Multiaddr,
+        cancel: CancellationToken,
+        signers_allowlist: Vec<OperatorPubKey>,
+    ) -> anyhow::Result<Self> {
+        let config = Configuration {
+            keypair: keypair.clone(),
+            idle_connection_timeout: Duration::from_secs(30),
+            listening_addr: local_addr,
+            allowlist,
+            connect_to,
+            signers_allowlist,
+            num_threads: Some(DEFAULT_NUM_THREADS),
+        };
+        let p2p_config = P2PConfig {
+            keypair: config.keypair.clone(),
+            idle_connection_timeout: config.idle_connection_timeout,
+            listening_addr: config.listening_addr.clone(),
+            allowlist: config.allowlist.clone(),
+            connect_to: config.connect_to.clone(),
+            signers_allowlist: config.signers_allowlist.clone(),
+        };
+
+        info!("creating an in-memory sled database");
+        let db = sled::Config::new().temporary(true).open()?;
+        let pool = ThreadPool::new(1);
+        let db = AsyncDB::new(pool, Arc::new(db));
+
+        info!("creating a swarm");
+        // let swarm = swarm::with_tcp_transport(&p2p_config)?;
```

Comment:
Tests with TCP transport are really hard.
Even the `libp2p` uses Memory Addresses: https://github.com/libp2p/rust-libp2p/blob/9228dc203f1173247089135cb0bc3c4ea7a1b676/swarm/tests/listener.rs#L23